proposals/1406_multiple-idps/README.md

```diff
@@ -0,0 +1,493 @@
+  resolves any possibility of conflict if the first IDP does not allow usernames to start with "admin/", for example
+  if usernames in that IDP are not allowed to contain a "/" character.
+
+Transformation expressions on the FederationDomain can be easily used to avoid identity collisions as desired.
+For example, the CEL expressions to prefix every username and group name are `"my-prefix:" + username` and
+`groups.map(g, "my-prefix:" + g)`.
+
+#### Upgrades
+
+Any upgrades into a new version of Pinniped which allows multiple IDPs will have a similar configuration. There will
+be a FederationDomain with no IDPs listed on the FederationDomain (since this was not previously allowed), and there
+will be only a single IDP CRD created in the namespace. Any other number of IDP CRDs previously resulted in an
+unusable Pinniped installation.
+
+During an upgrade, an existing installation of the Supervisor would already have a FederationDomain CR defined without
+an "identityProviders" section. To enable smooth upgrades, the "identityProviders" section would be optional.
+
+- The Supervisor code already correctly handles the case when there are no identity provider CRs defined. No users can
+  log in using that FederationDomain.
+- To handle the case where there is exactly one identity provider CR defined, the controller could load that CR for use
+  in the FederationDomain. The "displayName" of the identity provider would be automatically configured to be the same
+  name as the CR. This allows old configurations to continue working after upgrade.
+- When there are multiple identity provider CRs defined, the controller can fail to load the FederationDomain and update
+  its status to include an error saying that using a FederationDomain when multiple identity provider CRs are created
+  requires using the "identityProviders" field on the FederationDomain. This handles the case where the
+  user adds multiple identity provider CRs after upgrading, but forgets to add the "identityProviders" field to the
+  FederationDomain.
+
+If an admin adds "identityProviders" to a pre-existing FederationDomain and changes the "displayName" of a pre-existing
+identity provider, then:
+1. Pre-existing user sessions would fail to refresh, causing those users to need to interactively log in again, since
+   the identity provider names and types are already stored in user sessions for use during refreshes. This code already
+   has sufficient protections to ensure that we can never accidentally use a different identity provider during refresh
+   compared to which was used during initial login, even if there is an accidental name collision (via UID comparisons).
+2. Pre-existing kubeconfigs would now refer to the wrong identity provider name, and would need to be regenerated.
+
+If an admin wants to add a pre-existing identityProvider to a pre-existing FederationDomain without interrupting
```

Comment:
I'd be concerned about potential issues with any fallback approach where the user might be using an IDP which is not the IDP that they meant to use.

However, the `aliases` approach seems like it could work if we are concerned enough about this use case, and seems pretty elegant. The `aliases` and `displayNames` could be validated to ensure that no two `identityProviders` in the list are allowed to choose the same name/alias.

Let's consider how often this use case would impact users. This would be a one-time upgrade concern. Once an admin upgrades and reconfigures their FederationDomain to explicitly list `identityProviders`, then this concern will never arise again for that Supervisor for any future upgrades, and the aliases feature may not be useful to them anymore. Unless you maybe had another use case in mind for aliases in addition to the upgrade scenario?

If it is only for this one-time upgrade use case, then maybe its not worth the extra effort required to implement and maintain it?

Reply:
I think we should consider the initial upgrade from one IDP to multiple IDPs as well as general ease of ongoing use. Aliases would help with fixing typos in the `DisplayName`, updating the `DisplayName` for clarity, etc. Plus I think it really gives a way to make existing Kubeconfigs forward-compatible, which really helps illustrate that supervisor admins can use a human-readable `DisplayName`, and our docs can call that out.

In any case, I definitely agree that explicit aliases are much preferred over implicit behavior. Also that way we can check the `FederationDomain` for validity when it is read into memory on the primary Supervisor instance.

Reply:
This is a cool idea, but I think we should defer it and not include in the first version.